README.rst
.. image:: https://readthedocs.org/projects/zopeappapidoc/badge/?version=latest
        :target: https://zopeappapidoc.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://github.com/zopefoundation/zope.app.apidoc/actions/workflows/tests.yml/badge.svg
        :target: https://github.com/zopefoundation/zope.app.apidoc/actions/workflows/tests.yml
        :alt: Build Status

.. image:: https://coveralls.io/repos/github/zopefoundation/zope.app.apidoc/badge.svg?branch=master
        :target: https://coveralls.io/github/zopefoundation/zope.app.apidoc?branch=master
        :alt: Code Coverage

This Zope 3 package provides fully dynamic API documentation of Zope 3 and
registered add-on components. The package is very extensible and can be easily
extended by implementing new modules.

Documentation is available at  https://zopeappapidoc.readthedocs.io/


CHANGES.rst
=========
 CHANGES
=========

5.1 (unreleased)
================

- Nothing changed yet.


5.0 (2023-07-06)
================

- Drop support for Python 2.7, 3.5, 3.6.

- Add support for Python 3.11.


4.3.0 (2021-12-15)
==================

- Add support for Python 3.8, 3.9 and 3.10.

- Drop support for Python 3.4.


4.2.0 (2018-08-21)
==================

- Add support for Python 3.7.

- The root ``Code`` documentation node no longer allows incidental
  traversal and documentation of unregistered root modules such as
  ``re`` and ``logging`` (``builtins`` is special cased). These were
  not listed in the tables of contents or menus, and mostly served to
  slow down static exports. To document a root module, explicitly
  include it in ZCML with ``<apidoc:rootModule module="MODULE" />``.
  See `issue #20
  <https://github.com/zopefoundation/zope.app.apidoc/issues/20>`_.

- Fix ``codemodule.Module`` for modules that have a ``__file__`` of
  ``None``. This can be the case with namespace packages, especially
  under Python 3.7. See `issue #17 <https://github.com/zopefoundation/zope.app.apidoc/issues/17>`_.

- Rendering documentation for a class that has a ``__doc__`` property
  no longer fails but produces a descriptive message. See `issue 16
  <https://github.com/zopefoundation/zope.app.apidoc/issues/16>`_.

- Host documentation at https://zopeappapidoc.readthedocs.io/

- Add argument to ``static-apidoc`` for loading a specific ZCML file. To use this feature, the ZCML file you specify needs to
  establish a working Zope 3 publication environment. The easiest way to do so is to include this line in the ZCML:
  ``<include package='zope.app.apidoc' file='static.zcml' condition='have static-apidoc' />``.
  See `PR #13
  <https://github.com/zopefoundation/zope.app.apidoc/pull/13/>`_.

- Class Finder entries in live apidoc are now displayed on separate lines, like in static exports.
  See `PR #14 <https://github.com/zopefoundation/zope.app.apidoc/pull/14/>`_.

- Class Finder search in static exports will search on Enter, not just when clicking "Find".
  See `PR #15 <https://github.com/zopefoundation/zope.app.apidoc/pull/15/>`_.

- ``__main__.py`` files are no longer imported by the code documentation module.
  See `issue #22
  <https://github.com/zopefoundation/zope.app.apidoc/issues/22>`_.

- Cython functions registered as adapters on Python 2 no longer break
  page generation with an ``AttributeError``. See `issue 25
  <https://github.com/zopefoundation/zope.app.apidoc/issues/25>`_.

- Static exports no longer highlight lines in ZCML files. See `issue #24 
  <https://github.com/zopefoundation/zope.app.apidoc/issues/24>`_.

4.0.0 (2017-05-25)
==================

- Add support for Python 3.4, 3.5, 3.6 and PyPy.

- The long-deprecated layer configuration was removed. It was only
  ever available if the ``deprecatedlayers`` ZCML feature was installed.

- Modernize some of the templates. ``zope.app.apidoc`` can now be used
  with Chameleon 3.2 via z3c.pt and z3c.ptcompat.

- Declared install dependency on ``zope.app.exception``.

- Docstrings are treated as UTF-8 on Python 2.

- Handle keyword only arguments and annotations in function signatures
  on Python 3.

- Change the default documentation format to ``restructuredtext`` for
  modules that do not specify a ``__docformat__``. Previously it was
  ``structuredtext`` (STX).

3.7.5 (2010-09-12)
==================

- Define ``__file__`` in doctests to make them pass under Python 2.4.

3.7.4 (2010-09-01)
==================

- Prefer the standard library's doctest module to the one from zope.testing.

- Remove unneeded dependencies zope.app.component and zope.app.container

3.7.3 (2010-07-14)
==================

- Apply refactoring from #153309.
- Fix LP bug 605057: ZCML links were no longer working (Guilherme Salgado)

3.7.2 (2010-03-07)
==================

- Adapted tests for Python2.4


3.7.1 (2010-01-05)
==================

- Updated tests to work with zope.publisher 3.12 (using ``zope.login``).

3.7.0 (2009-12-22)
==================

- Updated tests to work with latest ``zope.testing`` and use ``zope.browserpage`` in
  favor of ``zope.app.pagetemplate``.

3.6.8 (2009-11-18)
==================

- Updated the tests after moving ``IPossibleSite`` and ``ISite`` to
  ``zope.component``.

3.6.7 (2009-09-29)
==================

- Updated the tests after moving ``ITraverser`` back to ``zope.traversing``.

3.6.6 (2009-09-15)
==================

- Made the tests work again with the most recent Zope Toolkit KGS.

3.6.5 (2009-07-24)
==================

- Update documentation file in ``zope.site`` from ``README.txt`` to
  ``site.txt``.

3.6.4 (2009-07-23)
==================

- The ``IContained`` interface moved to ``zope.location.interfaces``. Make a
  test pass.

3.6.3 (2009-05-16)
==================

- Explicitly defined default views.

- Replace relative url links with absolute ones.

- Added ``z3c`` packages to the code browser.

- Made ``bin/static-apidoc`` principally working (publisher and
  webserver mode). There are still some files which are not correctly
  fetched.

3.6.2 (2009-03-17)
==================

- Adapt principal registry book chapter to a new place, as it was moved
  from zope.app.security to zope.principalregistry.

- Remove zcml slugs and old zpkg-related files.

3.6.1 (2009-02-04)
==================

- When a module provides an interface or has an __all__ attribute,
  use one of those for the module documentation.  Fixes LP #323375.

- Undid broken link to ``savepoint.txt`` caused in 3.6.0.  The latest
  version of the transaction package puts savepoint.txt in the ``tests``
  subpackage.

- Expanded the presentation of module documentation.

- Class documentation now includes constructor information.

3.6.0 (2009-01-31)
==================

- Use zope.container instead of zope.app.container.

- Use zope.site instead of zope.app.component and zope.app.folder (in
  at least a few places).


interface.py
##############################################################################
#
# Copyright (c) 2004 Zope Corporation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Interface Inspection Utilities

$Id: browser.py 29199 2005-02-17 22:38:55Z srichter $
"""
__docformat__ = 'restructuredtext'

from zope.interface import Interface, providedBy
from zope.interface.interfaces import IInterface, ISpecification
from zope.interface.interfaces import IElement, IAttribute, IMethod
from zope.schema.interfaces import IField

from zope.app.apidoc.utilities import getPythonPath, renderText


def getElements(iface, type=IElement):
    """Return a dictionary containing the elements in an interface.

    The type specifies whether we are looking for attributes or methods."""
    items = {}
    for name in iface:
        attr = iface[name]
        if type.providedBy(attr):
            items[name] = attr
    return items


def getFieldsInOrder(iface,
                     _itemsorter=lambda x, y: cmp(x[1].order, y[1].order)):
    """Return a list of (name, field) tuples in native interface order."""
    items = getElements(iface, IField).items()
    items.sort(_itemsorter)
    return items


def getAttributes(iface):
    """Returns a list of attributes specified in the interface."""
    return [(name, attr)
            for name, attr in getElements(iface, IAttribute).items()
            if not (IField.providedBy(attr) or IMethod.providedBy(attr))]


def getMethods(iface):
    """Returns a list of methods specified in the interface."""
    return getElements(iface, IMethod).items()


def getFields(iface):
    """Returns a list of fields specified in the interface."""
    return getFieldsInOrder(iface)


def getInterfaceTypes(iface):
    """Return a list of interface types that are specified for this
    interface.

    Note that you should only expect one type at a time.
    """
    types = list(providedBy(iface).flattened())
    # Remove interfaces provided by every interface instance
    types.remove(ISpecification)
    types.remove(IElement)
    types.remove(Interface)    
    # Remove interface provided by every interface type
    types.remove(IInterface)
    return types


def getFieldInterface(field):
    """Return the interface representing the field."""
    name = field.__class__.__name__
    field_iface = None
    ifaces = tuple(providedBy(field).flattened())
    for iface in ifaces:
        # All field interfaces implement `IField`. In case the name match
        # below does not work, use the first `IField`-based interface found
        if field_iface is None and iface.extends(IField):
            field_iface = iface

        # Usually fields have interfaces with the same name (with an 'I')
        if iface.getName() == 'I' + name:
            return iface

    # If not even a `IField`-based interface was found, return the first
    # interface of the implemented interfaces list.  
    return field_iface or ifaces[0]
    

def getAttributeInfoDictionary(attr, format='zope.source.rest'):
    """Return a page-template-friendly information dictionary."""
    return {'name': attr.getName(),
            'doc': renderText(attr.getDoc() or u'', format=format)}


def getMethodInfoDictionary(method, format='zope.source.rest'):
    """Return a page-template-friendly information dictionary."""
    return {'name': method.getName(),
            'signature': method.getSignatureString(),
            'doc': renderText(method.getDoc() or u'', format=format)}


def getFieldInfoDictionary(field, format='zope.source.rest'):
    """Return a page-template-friendly information dictionary."""

    info = {'name': field.getName(),
            'required': field.required,
            'required_string': field.required and u'required' or u'optional',
            'default': repr(field.default),
            'title': field.title}

    # Determine the interface of the field
    iface = getFieldInterface(field)
    info['iface'] = {'name': iface.getName(), 'id': getPythonPath(iface)}

    # Determine the field class
    class_ = field.__class__
    info['class'] = {'name': class_.__name__,
                     'path': getPythonPath(class_).replace('.', '/')}

    # Render the field description
    info['description'] = renderText(field.description or u'', format=format)

    return info
